README.md
<h2 align="center"><img src="https://raw.githubusercontent.com/kbmjj123/zd-vscode-chatgpt/main/images/iconWhite.png" height="64"><br>Ask ChatGPT</h2>
<p align="center"><strong>ChatGPT, GPT-3 and Codex conversations in Visual Studio Code</strong></p>

<p align="center">
    <a href="https://marketplace.visualstudio.com/items?itemName=kbmjj123.zd-vscode-chatgpt" alt="Marketplace version">
        <img src="https://img.shields.io/visual-studio-marketplace/v/kbmjj123.zd-vscode-chatgpt?color=orange&label=VS%20Code%20Marketplace" />
    </a>
    <a href="https://marketplace.visualstudio.com/items?itemName=kbmjj123.zd-vscode-chatgpt" alt="Marketplace download count">
        <img src="https://img.shields.io/visual-studio-marketplace/d/kbmjj123.zd-vscode-chatgpt?color=blueviolet&label=Downloads" />
    </a>
    <a href="https://github.com/kbmjj123/zd-vscode-chatgpt" alt="Github star count">
        <img src="https://img.shields.io/github/stars/kbmjj123/zd-vscode-chatgpt?color=blue&label=Github%20Stars" />
    </a>
</p>

## 📢 New features - Keybindings, new custom prompts and enable/disable prompts

- You can now use/customize keybindings for your prompts. (i.e. Use `cmd+k cmd+shift+1` to run `ChatGPT: Add Tests` on the selected code on MacOS)
- You can now enable/disable prompts that you wish to see in the editor's context menus.
- There are 2 new prompts that you can customize and enable if the currently available prompts are insufficient for your use case.

# ChatGPT as your copilot to level up your developer experience

- ⏳ Google's LaMDA, Bard integration is upcoming once the API is available.
- ➕ ChatGPT Plus support.
- 🔥 Streaming conversation support for both browser and API Key method. ChatGPT will type as they think.
- 📝 Create files/projects or fix your code with one click or with keyboard shortcuts.
- 🤖 Zero-Config setup. Simply login to OpenAI as usual. Or use OpenAI's official GPT3 APIs.
- ➡️ Export all your conversation history at once in Markdown format.

<img src="https://raw.githubusercontent.com/kbmjj123/zd-vscode-chatgpt/main/images/chatgpt-gif.gif">

# Zero Configuration with Browser or API Key setup Methods

1. [**Default**] Use OpenAI's official ChatGPT APIs - Use your personal/organizational API Keys. This method provides many parameters to customize your prompt. Check out the GPT3 settings. For more details: [GPT3 OpenAI API Key](#api-key-setup)
2. Autologin - Uses browser to ask questions to ChatGPT. Zero-Config Autologin lets the extension grab the required tokens automatically using `puppeteer`. The extension will use the browser behind the scenes, so you are not expected to receive 4xx errors while asking ChatGPT via extension unless there is OpenAI outages. [Autologin Setup (Default)](#browser-setup)

<img src="https://raw.githubusercontent.com/kbmjj123/zd-vscode-chatgpt/main/images/methods.png">

# API Key Setup

Get your API Key ready from here: [OpenAI](https://beta.openai.com/account/api-keys)

1. Click `Login` in your extension or ask any coding question by selecting a code fragment.
2. Once asked provide your API Key to the extension
   - [Optional] You could also store your API Key in your settings.json. However, it's highly discouraged due to security reasons.

# Browser Setup

> ❕ Please do not interrupt the autologin flow and do not close the browser. Let the extension log you in and the browser will be minimized automatically.

🚀 All you need to do is click Login or ask a question to get started!

1. Click on the ChatGPT extension icon on your sidebar and hit `Login` button.

2. A new browser window (Default is `Chrome` but you may override it with any Chromium-based browser) will open up redirected to https://chat.openai.com/. Wait till you see login page, and click on Login on your browser.

   - If you see a `ChatGPT is at capacity right now` message, the extension will refresh the page every 2 seconds until it's passed. If OpenAI systems are experiencing issues and the extension fails to see Login page after 10 retries, the browser will be automatically closed.

3. Solve captchas if you are prompted and continue.

4. After successfully logging in, the browser will be minimized.

🔑 Use Google, Microsoft or standard OpenAI credentials to login. The email/password will be used to autofill the login form depending on the authentication type you choose. **NOTE**: We don't recommend storing your password in your settings.json since the file is plain text. However, you can opt-in to use it to reduce the friction during logins.

ℹ️ You will need to have a browser open and be logged in at all times. If you close the browser or your VS-Code instance, you will be asked to login again in your next session.

📝 You can have the extension auto-fill the email address and/or password during logins. Update the extension settings with those information for quicker login. NOTE: We never store any of this information locally or remotely.

🤖 Below is a sample autologin flow. Simply login & keep your browser minimized for dialogues with ChatGPT:

<img src="https://raw.githubusercontent.com/kbmjj123/zd-vscode-chatgpt/main/images/chatgpt-autologin.gif">

---

## Override settings with any Chromium-based browser

1. To use `Edge`, go to this URL: `edge://version` and copy the executable path used by your Edge browser.
2. Override the chromium path by going to vs-code settings and search for `chatgpt:chromiumPath`. Paste the executable path. i.e. `C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe`
3. [Maybe required] Run `ChatGPT: Reset session` command to clear your previous browser selection. This is required only if you have previously authenticated using a different browser.

### Run on Linux or remote - SSH workspace

There are community members who are able to run the extension in WSL, Linux, remote-SSH workspaces. Please check these issues for instructions.

Credits to [@EzyDark](https://github.com/EzYDark)

- [How to get it work in remote workspace (Remote - SSH) #39](https://github.com/kbmjj123/zd-vscode-chatgpt/issues/39#issuecomment-1370272656)
- [How to get it work under WSL2 #25](https://github.com/kbmjj123/zd-vscode-chatgpt/issues/25#issuecomment-1374833026)

# Features

The extension comes with context menu commands, copy/move suggested code into editor with one-click, conversation window and customization options for OpenAI's ChatGPT prompts.

## ChatGPT conversation window in vs-code

### 🆕 Export all your conversation history withs one click

<img src="https://raw.githubusercontent.com/kbmjj123/zd-vscode-chatgpt/main/images/export-convo.png">

---

### Ad-hoc prompt prefixes for you to customize what you are asking ChatGPT

Customize what you are asking with the selected code. The extension will remember your prompt for subsequent questions.

<img src="https://raw.githubusercontent.com/kbmjj123/zd-vscode-chatgpt/main/images/adhoc.png">

---

### Automatic partial code response detection

The extension will detect if ChatGPT didn't complete code in their answer and it will suggest automatic continuation and combination of answers

<img src="https://raw.githubusercontent.com/kbmjj123/zd-vscode-chatgpt/main/images/continue-combine.png">

---

### 🍻 Optimized for dialogue

<img src="https://raw.githubusercontent.com/kbmjj123/zd-vscode-chatgpt/main/images/conversation-mode-2.png">

---

### Edit and resend a previous prompt

<img src="https://raw.githubusercontent.com/kbmjj123/zd-vscode-chatgpt/main/images/edit-resend.png">

---

### Copy or insert the code ChatGPT is suggesting right into your editor.

<img src="https://raw.githubusercontent.com/kbmjj123/zd-vscode-chatgpt/main/images/code-actions.png">

---

### Ask free-form text questions that will be listed in the conversation window. The conversation is kept in cache until vs-code instance is closed.

<img src="https://raw.githubusercontent.com/kbmjj123/zd-vscode-chatgpt/main/images/rust.png">

# Use defaults or customize your code prompts

- `ChatGPT: Ad-hoc prompt`: Ad-hoc custom prompt prefix for the selected code. Right click on a selected block of code, run command.
  - You will be asked to fill in your preferred custom prefix and the extension will remember that string for your subsequent ad-hoc queries.
- `ChatGPT: Add tests`: Write tests for you. Right click on a selected block of code, run command.
  - "default": "Implement tests for the following code",
  - "description": "The prompt prefix used for adding tests for the selected code"
- `ChatGPT: Find bugs`: Analyze and find bugs in your code. Right click on a selected block of code, run command.
  - "default": "Find problems with the following code",
  - "description": "The prompt prefix used for finding problems for the selected code"
- `ChatGPT: Optimize`: Add suggestions to your code to improve. Right click on a selected block of code, run command.
  - "default": "Optimize the following code",
  - "description": "The prompt prefix used for optimizing the selected code"
- `ChatGPT: Explain`: Explain the selected code. Right click on a selected block of code, run command.
  - "default": "Explain the following code",
  - "description": "The prompt prefix used for explaining the selected code"
- `ChatGPT: Add comments`: Add comments for the selected code. Right click on a selected block of code, run command.
  - "default": "Add comments for the following code",
  - "description": "The prompt prefix used for adding comments for the selected code"
- `ChatGPT: Custom prompt 1`: Your custom prompt 1. It's disabled by default, please set to a custom prompt and enable it if you prefer using customized prompt
  - "default": "",
- `ChatGPT: Custom prompt 2`: Your custom prompt 2. It's disabled by default, please set to a custom prompt and enable it if you prefer using customized prompt
  - "default": "",

## Other available commands

- `ChatGPT: Ask anything`: Free-form text questions within conversation window.
- `ChatGPT: Reset session`: Clears the current session and resets your connection with ChatGPT
- `ChatGPT: Clear conversation`: Clears the conversation window and resets the thread to start a new conversation with ChatGPT.
- `ChatGPT: Export conversation`: Exports the whole conversation in Markdown for you to easily store and find the Q&A list.

## Customization settings

- Use proxy with autologin puppeteer setup
- Opt-in to use automation to authenticate OpenAI.
- You can configure the commands to use any prompts for the selected code!
- Opt-in to receive notification when ChatGPT sends you a message!

### Using Proxy

The autologin supports setting a proxy server. This is useful if you're running into rate limiting issues or if you want to use a proxy to hide your IP address. Right now this setting only supports http proxies. Don't provide a protocol in the setting.

To use a proxy, update the settings with your proxy server details. For more information on the format, see [here](https://www.chromium.org/developers/design-documents/network-settings).

Format examples:

- Authenticated: `myUsername:myPassword@my.proxy.com:3001`
- Anonymous: `204.137.172.37:999`

# Troubleshooting

- It's possible that OpenAI systems may experience issues responding to your queries due to high-traffic from time to time.
- If you get `404 NotFound` error, it means one of the parameters you provided is unknown (i.e. `chatgpt.gpt3.model`). Most likely switching to default `model` in your settings would fix this issue.
- If you get `400 BadRequest` error with API Key based method, it means that your conversation's length is more than GPT/Codex models can handle. Clear your conversation history with `ChatGPT: Clear conversation` command and retry sending your prompt.
- If you get `ChatGPT is at capacity right now` during autologin, the extension will refresh the page every 2 seconds until you see the login page. Refreshing may help in some cases, when there is a queue. This is unfortunately out of this extension's control. If, even after refresh, OpenAI shows capacity error, the browser will close automatically.
- If you get `ChatGPTAPI error 429`, it means that you are making Too Many Requests. Please wait and try again in a few moments. If it persists, restart your vs-code.
  - This could also be due to multiple browser/requests being active on OpenAI. Make sure that none of your API Keys are being actively used at the moment.
  - You could also try re-generating a new API Key [here](https://beta.openai.com/account/api-keys)
- If you see `ChatGPTAPI error terminated`, your requests are being throttled. Please try again later.
- If you encounter persistent issues with your queries
  - Try `ChatGPT: Reset session` command
  - As a last resort try restarting your VS-Code and retry logging in.
- If you are using Remote Development and cant login to ChatGPT
  - In settings.json add
  - "remote.extensionKind": {"kbmjj123.zd-vscode-chatgpt": ["ui"]}

# Disclaimer and Credits

- This public repository is only used for documentation purposes at the moment. It's due to various reasons outlined [here](https://github.com/kbmjj123/zd-vscode-chatgpt/issues/68). Please make sure you are comfortable with the disclaimers below before using the extension.
- There is no guarantee that the extension will continue to work as-is without any issues or side-effects. Please use it at your own risk. It may stop working without a notice e.g. OpenAI may decide to change all or some of its functionality, which will affect this extension.
- This extension never uses/stores your personally identifiable information. There are some optional settings that you may opt-in to use i.e. OpenAI API Key. Please be careful of what you are storing in your settings.json file, since vs-code may sync them across their instances and it's outside of this extension's boundary. We recommend not storing any personally identifiable information in your settings.json.
- This extension collects metadata to improve its features. No personally identifiable information is collected. You can opt-out from telemetry either by setting the global 'telemetry.telemetryLevel' or 'chatgpt.telemetry.disable'. The extension will respect both of these settings and will collect metadata only if both allow telemetry. We use the official telemetry package provided by the vscode team [here](https://github.com/Microsoft/vscode-extension-telemetry) to understand this extension's usage patterns better to plan new feature releases i.e. popularity of Browser-based autologin vs. API Key method setting.
- We assume no responsibility of any issues that you may face using this extension. Your use of OpenAI services is subject to OpenAI's [Privacy Policy](https://openai.com/privacy/) and [Terms of Use](https://openai.com/terms/).
- 💻 Open AI ChatGPT: https://chat.openai.com/
- 🖼️ Open AI Dall-E-2: https://openai.com/dall-e-2/
- 🧪 Uses [unofficial nodejs OpenAI API wrapper](https://github.com/transitive-bullshit/chatgpt-api) for conversational fetch calls to api.openai.com.


CHANGELOG.md
# Change Log

All notable changes to the "zd-vscode-chatgpt" extension will be documented in this file.

## [3.6.0] - 2022-02-12

- Keybindings, new custom prompts and enable/disable prompts
- You can now use/customize keybindings for your prompts. (i.e. Use `cmd+k cmd+shift+1` to run `ChatGPT: Add Tests` on the selected code on MacOS)
- You can now enable/disable prompts that you wish to see in the editor's context menus.
- There are 2 new prompts that you can customize and enable if the currently available prompts are insufficient for your use case.

## [3.5.5] - 2022-02-05

- ChatGPT Plus support. If you have a Plus subscription, you can now use the \*-paid models for your prompts.

## [3.5.0] - 2022-02-04

- The pre-release Official ChatGPT model is available via OpenAI API Key method.

## [3.4.0] - 2022-01-20

- Streaming conversation is now available. You will see results as ChatGPT / Codex / GPT models are typing.

## [3.2.0] - 2022-01-15

- You can now use `Google` or `Microsoft` Authentication types during login flow to auto-fill the forms with the email/password you provide the extension with.

## [3.1.0] - 2022-01-12

- You can now have conversations with GPT3 using the Official OpenAI APIs.
- The GPT3/Codex dialogue support is similar to what ChatGPT offers. However, note that ChatGPT is more advanced in dialogues.

## [3.0.0] - 2022-01-06

- Added support for Official OpenAI APIs for GPT3 and Codex using API Keys.

## [2.6.0] - 2022-01-04

- Added proxy server. Addressing issue #32
- Added custom timeout. Addressing issue #33

## [2.3.0] - 2022-12-27

- You can now add custom prefix commands for the selected code. The extension will remember the prompt for subsequent queries. Addressing issue #28

## [2.2.0] - 2022-12-27

- You can now use any Chromium-based browser for autologin. Search for setting `chatgpt:chromiumpath` to override it. Check out the Readme for more.

## [2.1.0] - 2022-12-26

- Automatically login without needing to copy-paste session cookies is now live and enabled by default
- UX improvements: moved `Clear conversation` and `Export to MD` buttons to flyout.
- Improved error handling
- Misc. bug fixes

## [1.8.0] - 2022-12-14

- 🚀 [Experimental Feature] Automatically login without needing to copy-paste session cookies

## [1.7.0] - 2022-12-12

- 🚀 Automatically login without needing to copy-paste session cookies
- OpenAI servers are experiencing exceptionally high demand, which may cause intermittent errors in this extension.

## [1.6.0] - 2022-12-11

- 🆕 New vs-code commands to Clear and Export conversations

## [1.5.0] - 2022-12-11

- 🆕 Clear conversations and reset thread
- 🆕 Export all your conversation dialogue into a new tab with single click
- 🆕 Added a new introduction page that shows up before the conversation starts

## [1.4.3] - 2022-12-10

- 🚀 Better theming support in markdown

## [1.4.2] - 2022-12-10

- Update icons and SVGs used in conversation window.
- Show copy to clipboard buttons on hover for better UX.

## [1.4.0] - 2022-12-09

- Introduced edit and resend feature 🚀. Addressing #6
- Edit previous prompt in-place
- Better font-family support for code highlighter

## [1.3.0] - 2022-12-08

- Continuous conversations with ChatGPT with follow-ups. Addressing #7
- UX improvements

## [1.2.0] - 2022-12-07

- Automatically highlight code in conversation view
- UX improvements addressing #4 and #8

## [1.1.0] - 2022-12-07

- Adding a new configuration to opt-in to receive notifications when chatgpt responds.
  - This config is set to false by default to prevent notification spams.
  - If users are interested in it due to high-volume latencies, they can opt-in to get notifications when answers are ready.

## [1.0.0] - 2022-12-07

- Introducing new ways to interact with the suggested code
- Copy code to clipboard
- Insert code to text editor
- Create a new file with the ChatGPT suggested code

## [0.1.0] - 2022-12-6

- Initial release of the ChatGPT - VS Code extension.


src/extension.ts
/**
 * @author Ali Gençay
 * https://github.com/kbmjj123/zd-vscode-chatgpt
 *
 * @license
 * Copyright (c) 2022 - Present, Ali Gençay
 *
 * All rights reserved. Code licensed under the MIT license
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 */

import * as vscode from "vscode";
import ChatGptViewProvider from './chatgpt-view-provider';

export async function activate(context: vscode.ExtensionContext) {
    const chatGptExtensionConfig = vscode.workspace.getConfiguration("chatgpt");
    const provider = new ChatGptViewProvider(context);

    const view = vscode.window.registerWebviewViewProvider(
        "zd-vscode-chatgpt.view",
        provider,
        {
            webviewOptions: {
                retainContextWhenHidden: true,
            },
        }
    );

    const freeText = vscode.commands.registerCommand("zd-vscode-chatgpt.freeText", async () => {
        const value = await vscode.window.showInputBox({
            prompt: "Ask anything...",
        });

        if (value) {
            provider?.sendApiRequest(value);
        }
    });

    const resetThread = vscode.commands.registerCommand("zd-vscode-chatgpt.clearConversation", async () => {
        provider?.sendMessage({ type: 'clearConversation' }, true);
    });

    const exportConversation = vscode.commands.registerCommand("zd-vscode-chatgpt.exportConversation", async () => {
        provider?.sendMessage({ type: 'exportConversation' }, true);
    });

    const clear = vscode.commands.registerCommand("zd-vscode-chatgpt.clearSession", () => {
        context.globalState.update("chatgpt-session-token", null);
    });

    const configChanged = vscode.workspace.onDidChangeConfiguration(e => {
        if (e.affectsConfiguration('chatgpt.response.showNotification')) {
            provider.subscribeToResponse = vscode.workspace.getConfiguration("chatgpt").get("response.showNotification") || false;
        }
    });

    const commands = [
        ["zd-vscode-chatgpt.addTests", "promptPrefix.addTests"],
        ["zd-vscode-chatgpt.findProblems", "promptPrefix.findProblems"],
        ["zd-vscode-chatgpt.optimize", "promptPrefix.optimize"],
        ["zd-vscode-chatgpt.explain", "promptPrefix.explain"],
    ];

    const registeredCommands = commands.map(([command, configKey]) =>
        vscode.commands.registerCommand(command, () => {
            const commandPrefix = chatGptExtensionConfig.get(configKey) as string;
            const editor = vscode.window.activeTextEditor;
            if (!editor) {
                return;
            }

            const selection = editor.document.getText(editor.selection);
            if (selection) {
                provider?.sendApiRequest(commandPrefix, selection);
            }
        })
    );

    context.subscriptions.push(view, freeText, resetThread, exportConversation, clear, configChanged, ...registeredCommands);
}

export function deactivate() { }


src/chatgpt-view-provider.ts
/**
 * @author Ali Gençay
 * https://github.com/kbmjj123/zd-vscode-chatgpt
 *
 * @license
 * Copyright (c) 2022 - Present, Ali Gençay
 *
 * All rights reserved. Code licensed under the MIT license
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 */

import { ChatGPTAPI, ChatGPTConversation } from 'chatgpt';
import * as vscode from 'vscode';

export default class ChatGptViewProvider implements vscode.WebviewViewProvider {
    private webView?: vscode.WebviewView;
    private chatGptApi?: ChatGPTAPI;
    private chatGptConversation?: ChatGPTConversation;
    private sessionToken?: string;
    public subscribeToResponse: boolean;

    /**
     * Message to be rendered lazily if they haven't been rendered
     * in time before resolveWebviewView is called.
     */
    private leftOverMessage?: any;

    constructor(private context: vscode.ExtensionContext) {
        this.subscribeToResponse = vscode.workspace.getConfiguration("chatgpt").get("response.showNotification") || false;
    }

    public resolveWebviewView(
        webviewView: vscode.WebviewView,
        _context: vscode.WebviewViewResolveContext,
        _token: vscode.CancellationToken,
    ) {
        this.webView = webviewView;

        webviewView.webview.options = {
            // Allow scripts in the webview
            enableScripts: true,

            localResourceRoots: [
                this.context.extensionUri
            ]
        };

        webviewView.webview.html = this.getWebviewHtml(webviewView.webview);

        webviewView.webview.onDidReceiveMessage(async data => {
            switch (data.type) {
                case 'addFreeTextQuestion':
                    this.sendApiRequest(data.value);
                    break;
                case 'editCode':
                    vscode.window.activeTextEditor?.insertSnippet(new vscode.SnippetString(data.value));
                    break;
                case 'openNew':
                    const document = await vscode.workspace.openTextDocument({
                        content: data.value,
                        language: data.language
                    });
                    vscode.window.showTextDocument(document);
                    break;
                case 'clearConversation':
                    this.prepareConversation(true);
                    break;
                default:
                    break;
            }
        });

        if (this.leftOverMessage != null) {
            // If there were any messages that wasn't delivered, render after resolveWebView is called.
            this.sendMessage(this.leftOverMessage);
            this.leftOverMessage = null;
        }
    }

    private async prepareConversation(reset?: boolean): Promise<boolean> {
        this.sessionToken = await this.context.globalState.get("chatgpt-session-token") as string;

        if (this.sessionToken == null) {
            await vscode.window
                .showInputBox({
                    title: "OpenAPI ChatpGPT session token",
                    prompt: "Please enter your OpenAPI session token (__Secure-next-auth.session-token). See Readme for more details on how to get the session token",
                    ignoreFocusOut: true,
                    placeHolder: "Enter the JWT Token starting with ey***"
                })
                .then((value) => {
                    reset = true;
                    this.sessionToken = value!;
                    this.context.globalState.update("chatgpt-session-token", this.sessionToken);
                });
        }

        if (reset || this.chatGptApi == null || this.chatGptConversation == null) {
            try {
                this.chatGptApi = new ChatGPTAPI({ sessionToken: this.sessionToken });
                this.chatGptConversation = this.chatGptApi.getConversation();
            } catch (error: any) {
                vscode.window.showErrorMessage("Failed to instantiate the ChatGPT API. Try ChatGPT: Clear session", error?.message);
                this.sendMessage({ type: 'addError' });
                return false;
            }
        }

        return true;
    }

    public async sendApiRequest(prompt: string, code?: string) {
        await this.prepareConversation();

        if (this.chatGptApi == null || this.chatGptConversation == null) {
            return;
        }

        let response: string;
        let question = prompt;

        if (code != null) {
            // Add prompt prefix to the code if there was a code block selected
            question = `${prompt}: ${code}`;
        }

        // If the ChatGPT view is not in focus/visible; focus on it to render Q&A
        if (this.webView == null) {
            await vscode.commands.executeCommand('zd-vscode-chatgpt.view.focus');
        } else {
            this.webView?.show?.(true);
        }

        this.sendMessage({ type: 'addQuestion', value: prompt, code });

        try {
            await this.chatGptApi.ensureAuth();

            response = await this.chatGptConversation.sendMessage(question, {
                timeoutMs: 2 * 60 * 1000
            });

            if (this.subscribeToResponse) {
                vscode.window.showInformationMessage("ChatGPT responded to your question.", "Open conversation").then(async () => {
                    await vscode.commands.executeCommand('zd-vscode-chatgpt.view.focus');
                });
            }
        } catch (error: any) {
            vscode.window.showErrorMessage("An error occured. If the issue persists try 'ChatGPT: Clear session.'", error?.message);
            this.sendMessage({ type: 'addError' });
            return;
        }

        this.sendMessage({ type: 'addResponse', value: response });
    }

    /**
     * Message sender, stores if a message cannot be delivered
     * @param message Message to be sent to WebView
     * @param ignoreMessageIfNullWebView We will ignore the command if webView is null/not-focused
     */
    public sendMessage(message: any, ignoreMessageIfNullWebView?: boolean) {
        if (this.webView) {
            this.webView?.webview.postMessage(message);
        } else if (!ignoreMessageIfNullWebView) {
            this.leftOverMessage = message;
        }
    }

    private getWebviewHtml(webview: vscode.Webview) {
        const scriptUri = webview.asWebviewUri(vscode.Uri.joinPath(this.context.extensionUri, 'media', 'main.js'));
        const stylesMainUri = webview.asWebviewUri(vscode.Uri.joinPath(this.context.extensionUri, 'media', 'main.css'));

        const vendorHighlightCss = webview.asWebviewUri(vscode.Uri.joinPath(this.context.extensionUri, 'media', 'vendor', 'highlight.min.css'));
        const vendorHighlightJs = webview.asWebviewUri(vscode.Uri.joinPath(this.context.extensionUri, 'media', 'vendor', 'highlight.min.js'));
        const vendorMarkedJs = webview.asWebviewUri(vscode.Uri.joinPath(this.context.extensionUri, 'media', 'vendor', 'marked.min.js'));
        const vendorTailwindJs = webview.asWebviewUri(vscode.Uri.joinPath(this.context.extensionUri, 'media', 'vendor', 'tailwindcss.3.2.4.min.js'));
        const vendorTurndownJs = webview.asWebviewUri(vscode.Uri.joinPath(this.context.extensionUri, 'media', 'vendor', 'turndown.js'));

        return `<!DOCTYPE html>
			<html lang="en">
			<head>
				<meta charset="UTF-8">
				<meta name="viewport" content="width=device-width, initial-scale=1.0">

				<link href="${stylesMainUri}" rel="stylesheet">
				<link href="${vendorHighlightCss}" rel="stylesheet">
				<script src="${vendorHighlightJs}"></script>
				<script src="${vendorMarkedJs}"></script>
				<script src="${vendorTailwindJs}"></script>
				<script src="${vendorTurndownJs}"></script>
			</head>
			<body class="overflow-hidden">
				<div class="flex flex-col h-screen">
					<div id="introduction" class="flex h-full items-center justify-center px-6 w-full relative">
						<div class="flex items-start text-center gap-3.5">
							<div class="flex flex-col gap-3.5 flex-1">
								<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" aria-hidden="true" class="w-6 h-6 m-auto">
									<path stroke-linecap="round" stroke-linejoin="round" d="M3.75 13.5l10.5-11.25L12 10.5h8.25L9.75 21.75 12 13.5H3.75z"></path>
								</svg>
								<h2 class="text-lg font-normal">Features</h2>
								<ul class="flex flex-col gap-3.5">
									<li class="w-full bg-gray-50 dark:bg-white/5 p-3 rounded-md">Optimized for dialogue</li>
									<li class="w-full bg-gray-50 dark:bg-white/5 p-3 rounded-md">Improve your code, add tests & find bugs</li>
									<li class="w-full bg-gray-50 dark:bg-white/5 p-3 rounded-md">Copy or create new files automatically</li>
									<li class="w-full bg-gray-50 dark:bg-white/5 p-3 rounded-md">Syntax highlighting with auto language detection</li>
								</ul>
							</div>
							<div class="flex flex-col gap-3.5 flex-1">
								<svg stroke="currentColor" fill="none" stroke-width="1.5" viewBox="0 0 24 24" stroke-linecap="round" stroke-linejoin="round" class="w-6 h-6 m-auto" height="1em" width="1em"
									xmlns="http://www.w3.org/2000/svg">
									<path d="M10.29 3.86L1.82 18a2 2 0 0 0 1.71 3h16.94a2 2 0 0 0 1.71-3L13.71 3.86a2 2 0 0 0-3.42 0z"></path>
									<line x1="12" y1="9" x2="12" y2="13"></line>
									<line x1="12" y1="17" x2="12.01" y2="17"></line>
								</svg>
								<h2 class="text-lg font-normal">Limitations</h2>
								<ul class="flex flex-col gap-3.5">
									<li class="w-full bg-gray-50 dark:bg-white/5 p-3 rounded-md">May occasionally take long time to respond/fail</li>
									<li class="w-full bg-gray-50 dark:bg-white/5 p-3 rounded-md">May throw HTTP 429, if you make too many requests</li>
									<li class="w-full bg-gray-50 dark:bg-white/5 p-3 rounded-md">If issues persist, clear your session and re-login</li>
								</ul>
							</div>
						</div>
						<p class="absolute bottom-0 max-w-sm text-center text-xs text-slate-500">Get your session token <a href="https://chat.openai.com">here</a>.<br /><a href="https://github.com/kbmjj123/zd-vscode-chatgpt">©️ Open source</a></p>
					</div>

					<div class="flex-1 overflow-y-auto" id="qa-list"></div>

					<div id="in-progress" class="pl-4 pt-2 flex items-center hidden">
						<div class="typing">Typing</div>
						<div class="spinner">
							<div class="bounce1"></div>
							<div class="bounce2"></div>
							<div class="bounce3"></div>
						</div>
					</div>

					<div id="chat-button-wrapper" class="w-full flex gap-4 justify-center items-center mt-2 hidden">
						<button class="flex gap-2 justify-center items-center rounded-lg p-2" id="clear-button">
							<svg stroke="currentColor" fill="none" stroke-width="2" viewBox="0 0 24 24" stroke-linecap="round" stroke-linejoin="round" class="w-4 h-4" xmlns="http://www.w3.org/2000/svg"><polyline points="1 4 1 10 7 10"></polyline><polyline points="23 20 23 14 17 14"></polyline><path d="M20.49 9A9 9 0 0 0 5.64 5.64L1 10m22 4l-4.64 4.36A9 9 0 0 1 3.51 15"></path></svg>
							Clear conversation
						</button>
						<button class="flex gap-2 justify-center items-center rounded-lg p-2" id="export-button">
							<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" class="w-4 h-4">
								<path stroke-linecap="round" stroke-linejoin="round" d="M3 16.5v2.25A2.25 2.25 0 005.25 21h13.5A2.25 2.25 0 0021 18.75V16.5M16.5 12L12 16.5m0 0L7.5 12m4.5 4.5V3" />
							</svg>
							Export all
						</button>
					</div>

					<div class="p-4 flex items-center pt-2">
						<div class="flex-1 textarea-wrapper">
							<textarea
								type="text"
								rows="1"
								id="question-input"
								placeholder="Ask a question..."
								onInput="this.parentNode.dataset.replicatedValue = this.value"></textarea>
						</div>
						<button title="Submit prompt" class="right-8 absolute ask-button rounded-lg p-0.5 ml-5" id="ask-button">
							<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" class="w-6 h-6"><path stroke-linecap="round" stroke-linejoin="round" d="M6 12L3.269 3.126A59.768 59.768 0 0121.485 12 59.77 59.77 0 013.27 20.876L5.999 12zm0 0h7.5" /></svg>
						</button>
					</div>
				</div>

				<script src="${scriptUri}"></script>
			</body>
			</html>`;
    }
}
